lang/eval.go

```diff
@@ -251,20 +291,15 @@ func (s *Scope) evalContext(refs []*addrs.Reference, selfAddr addrs.Referenceabl
 				panic(fmt.Errorf("unsupported ResourceMode %s", subj.Resource.Mode))
 			}
 
-			val, valDiags := normalizeRefValue(s.Data.GetResourceInstance(subj, rng))
+			val, valDiags := normalizeRefValue(s.Data.GetResource(subj.ContainingResource(), rng))
```

Comment:
Could we remove this `case` by putting something like the following fixup before the `switch` here:

```go
if addr, ok := rawSubj.(addrs.ResourceInstance); ok {
    rawSubj = addr.ContainingResource()
}
```

I think that wouldn't have been possible with how `self` was handled before, where each of these cases was responsible for detecting if it was `self` and setting it, but now that `self` is being handled as special anyway it seems like that removes the remaining need for this clause. I'm probably missing something though, since there seem to be a bunch of different codepaths through there and I'm not sure I'm catching all of the interactions.

Reply:
Good call. I think I meant to try and coalesce these cases and forgot